decimal_test.go

```diff
@@ -743,3 +744,100 @@ func TestSizeof(t *testing.T) {
 		t.Errorf("sizeof(Context) changed: %d", s)
 	}
 }
+
+func TestUnmarshalText(t *testing.T) {
+	tests := []struct {
+		s         string
+		expectedD *Decimal
+		expectedE string
+	}{
+		{s: "1.1", expectedD: New(11, -1)},
+		{s: "null", expectedE: "parse mantissa: null"},
+		{s: "", expectedE: "parse mantissa: "},
+		{s: "foo", expectedE: "parse mantissa: foo"},
+		{s: "1.1.1", expectedE: "parse mantissa: 11.1"},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.s, func(t *testing.T) {
+			d := Decimal{}
+			err := d.UnmarshalText([]byte(tc.s))
+			if err == nil {
+				if tc.expectedE != "" {
+					t.Errorf("Expected error %s but got no error", tc.expectedE)
+				}
+				if d.Cmp(tc.expectedD) != 0 {
+					t.Errorf("Failed to correctly unmarshal JSON: %v != %v", d, tc.expectedD)
+				}
+			} else {
+				if tc.expectedE == "" {
+					t.Error("Got unexpected error:", err)
+				} else if err.Error() != tc.expectedE {
+					t.Errorf("Expected error %s but got %s instead", tc.expectedE, err.Error())
+				}
+			}
+		})
+	}
+}
+
+func TestMarshalText(t *testing.T) {
+	tests := []struct {
+		d         *Decimal
+		expectedS string
+	}{
+		{d: New(11, -1), expectedS: "1.1"},
+		{d: New(11, -10), expectedS: "1.1E-9"},
+		{d: New(11, 10), expectedS: "1.1E+11"},
+		{d: nil, expectedS: "<nil>"},
+	}
+
+	for _, tc := range tests {
+		subTestName := "nil"
+		if tc.d != nil {
+			subTestName = tc.d.String()
+		}
+		t.Run(subTestName, func(t *testing.T) {
+			b, err := tc.d.MarshalText()
+			if err != nil {
+				t.Fatal("Unexpected error marshaling to text:", err)
+			}
+			s := string(b)
+			if s != tc.expectedS {
+				t.Errorf("MarshalText() test failed. %s != %s", s, tc.expectedS)
+			}
+		})
+	}
+}
+
+var encodingTests = []string{
```

Comment:
This can be moved inside TestJSONEncoding since that's the only place it's used.